Libraries/Opc.Ua.Client/SessionReconnectHandler.cs

```diff
@@ -83,32 +120,103 @@ protected virtual void Dispose(bool disposing)
         /// <value>The session.</value>
         public ISession Session => m_session;
 
+        /// <summary>
+        /// The internal state of the reconnect handler.
+        /// </summary>
+        public ReconnectState State
+        {
+            get
+            {
+                lock (m_lock)
+                {
+                    if (m_reconnectTimer == null)
+                    {
+                        return ReconnectState.Disposed;
+                    }
+                    return m_state;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Cancel a reconnect in progress.
+        /// </summary>
+        public void CancelReconnect()
+        {
+            lock (m_lock)
+            {
+                if (m_reconnectTimer == null)
+                {
+                    return;
+                }
+
+                if (m_state == ReconnectState.Triggered)
+                {
+                    m_session = null;
+                    EnterReadyState();
+                    return;
+                }
+
+                m_cancelReconnect= true;
+            }
+        }
+
         /// <summary>
         /// Begins the reconnect process.
         /// </summary>
-        public void BeginReconnect(ISession session, int reconnectPeriod, EventHandler callback)
+        public ReconnectState BeginReconnect(ISession session, int reconnectPeriod, EventHandler callback)
         {
-            BeginReconnect(session, null, reconnectPeriod, callback);
+            return BeginReconnect(session, null, reconnectPeriod, callback);
         }
 
         /// <summary>
         /// Begins the reconnect process using a reverse connection.
         /// </summary>
-        public void BeginReconnect(ISession session, ReverseConnectManager reverseConnectManager, int reconnectPeriod, EventHandler callback)
+        public ReconnectState BeginReconnect(ISession session, ReverseConnectManager reverseConnectManager, int reconnectPeriod, EventHandler callback)
         {
             lock (m_lock)
             {
-                if (m_reconnectTimer != null)
+                if (m_reconnectTimer == null)
                 {
                     throw new ServiceResultException(StatusCodes.BadInvalidState);
                 }
 
-                m_session = session;
-                m_reconnectFailed = false;
+                // cancel reconnect requested, if possible
+                if (session == null)
+                {
+                    if (m_state == ReconnectState.Triggered)
+                    {
+                        m_session = null;
+                        EnterReadyState();
+                        return m_state;
+                    }
+                    // reconnect already in progress, schedule cancel
```

Comment:
Comment seems not to be aligned with code (if reconnect is in progress should be checked on m_state) and I also not understand why a 2nd call to BeginnReconnect should end up in cancelling